### DISCUSSION TITLE
Tenant-independent pages

### DISCUSSION THREAD

**Discussion by juliangums** [filamentphp/filament#13431] [Ideas]
One feature that has been requested a few times and that I struggle with not having are tenant-independent pages on panels that use tenancy. I need this feature either way but wanted to open a discussion about it first to define the specs in a way that benefit other too. 

My reasoning why this is needed has already been described in a previous discussion that has been closed by the creator without finding a valid answer: https://github.com/filamentphp/filament/discussions/11705
It also contains an implementation concept that I have successfully tested. Perhaps the problem statement wasn't clearly articulated by me, though. I was hoping it will get some more traction if I open a fresh discussion.

**Background:** When a user is not part of any tenant, they get shown the tenant registration page. While that might be perfect for some, there are use cases in which a user is not allowed to create a tenant. In that case, they are simply stuck in a 404 after logging in. Ideally, they should see something explaining them to contact their system admin or whatever. They should also have access to their profile for example as that does not depend on a tenant. There might be other pages and resources that don't depend on a tenant either. Perhaps they need to submit a ticket. There should be a way of marking a page/resource as tenant-dependent (default), tenant-independent or visible in both states. 

**Proposed workarounds:** A second panel is not an option. Routes are separated, making it not only messy but impossible to seamlessly integrate. It would have two separate logins or, if one is disabled, auth redirects for one panel point to the wrong route.

**Current solution:** As it needs changes to the routes and the route group blocks the entire namespace, there isn't an easy way to extend Filament to make this work. A fork is currently the only solution.

**Suggested solution:**
We could have two values for pages and resources `$showOnTenantDependentRoute` and `$showOnTenantIndependentRoute` as a boolean or closure and then add some logic to the routes to register them where they are needed. We then add some additional logic in the sidebar to not render the tenant menu if on a tenant-independent route. This worked fine when I tried. I'm looking forward for ideas to improve this.

@zepfietje @danharrin would you consider merging this feature?

**Comment by danharrin** [comment]
Why not make it easier to redirect between panels? The `$showOnTenantDependentRoute` property just sounds like a complicated way around using 2 panels which keeps the code separate and allows customizations based on which panel the user is in

**Reply by juliangums** [reply]
@danharrin in theory that'd be a solution, if auth redirection can be set so it directs to the other panel.
The only issue I would have with it is that those two panels can't operate in the same path. I could prefix the tenant-dependant one with 'team' for example. Definitely not a huge problem, just OCD as paths become unnecessarily long. But it would be an acceptable workaround if auth redirection of one panel can be set to the other. I'll open an issue for that, as you suggested.

**Comment by juliangums** [comment]
For anyone interested in how to achieve this right now in one panel without path/auth issues, I did find a much simpler working approach to make things work on the same panel by overriding the view and using the tenant registration class  to register our tenant-independent routes:

1) Depending on what layout you use, extend either of these files:
https://github.com/filamentphp/filament/blob/3.x/packages/panels/resources/views/components/sidebar/index.blade.php#L105
https://github.com/filamentphp/filament/blob/3.x/packages/panels/resources/views/components/topbar/index.blade.php#L60
You can do this by copying it to your resource folder like `resources/views/vendor/filament-panels/components/sidebar/index.blade.php`

2) Add a third condition inside the lines specified in the link above ` && ! is_null(filament()->getTenant())`.

3) Create a tenant registration class. Extend the Page class instead of the RegisterTenant class and add these three methods:
```php
<?php

namespace App\Filament\Pages;

use Filament\Pages\Page;
use Filament\Panel;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Facades\Route;

class NoTenant extends Page
{
    
    ...

    public static function getRelativeRouteName(): string
    {
        return 'registration';
    }

    public static function registerRoutes(Panel $panel): void
    {
        static::routes($panel);
    }

    public static function getUrl(array $parameters = [], bool $isAbsolute = true, ?string $panel = null, ?Model $tenant = null): string
    {
        return route(static::getRouteName($panel), $isAbsolute);
    }
}
```
This is just a simple example for one page as I simply needed a page to tell people to contact their system admin instead of showing a 404. But you can abuse the `registerRoutes()` method in this class to register as many pages as you like outside the tenant-scoped routes. You could also register resources. An example would be this, to make a profile page accessible:
```php
       Route::name('pages.')->group(function () use ($panel){
            Route::get('profile', MyProfilePage::class);
        });
```

4) Register the above class in the panel with `->tenantRegistration(NoTenant::class)`.

**Comment by mathieutu** [comment]
For those who need to register pages outside the tenant, here is something even simpler:

You can create standard pages, directly in your resources. If you don't want to register the standard routes, don't add them to the resources `getPages` function.

In your panel service provider, you can use the `routes()`, `authenticatedRoutes()` or `tenantRoutes()` methods to add any custom route you want: 

```php
return $panel
    ->authenticatedRoutes(function () {
        Route::get('contracts/{record}/sign', \App\Filament\Resources\ContractResource\Pages\SignContract::class)->name('contracts.sign');
    })
```

Depending on the view and layout you use, you'll have to update them in the page, to not use any tenant reference.
